rows.go

```diff
@@ -7,97 +7,160 @@ import "C"
-			dst[i] = (*[1 << 31]int8)(unsafe.Pointer(colData))[r.cursor]
-		case C.DUCKDB_TYPE_SMALLINT:
-			dst[i] = (*[1 << 31]int16)(unsafe.Pointer(colData))[r.cursor]
-		case C.DUCKDB_TYPE_INTEGER:
-			dst[i] = (*[1 << 31]int32)(unsafe.Pointer(colData))[r.cursor]
-		case C.DUCKDB_TYPE_BIGINT:
-			dst[i] = (*[1 << 31]int64)(unsafe.Pointer(colData))[r.cursor]
-		case C.DUCKDB_TYPE_HUGEINT: //int128...
-			var v = (*[1 << 31]HugeInt)(unsafe.Pointer(colData))[r.cursor]
-			if v, err := v.Int64(); err != nil {
-				return err
-			} else {
-				dst[i] = v
-			}
-		case C.DUCKDB_TYPE_FLOAT:
-			dst[i] = (*[1 << 31]float32)(unsafe.Pointer(colData))[r.cursor]
-		case C.DUCKDB_TYPE_DOUBLE:
-			dst[i] = (*[1 << 31]float64)(unsafe.Pointer(colData))[r.cursor]
-		case C.DUCKDB_TYPE_DATE:
-			val := (*[1 << 31]C.duckdb_date)(unsafe.Pointer(colData))[r.cursor]
-			dst[i] = time.UnixMilli(0).Add(time.Duration(int64(val.days)) * 24 * time.Hour)
-		case C.DUCKDB_TYPE_VARCHAR:
-			dst[i] = C.GoString((*[1 << 31]*C.char)(unsafe.Pointer(colData))[r.cursor])
-		case C.DUCKDB_TYPE_BLOB:
-			blob := C.duckdb_value_blob(r.res, C.idx_t(i), C.idx_t(r.cursor))
-			dst[i] = C.GoBytes(blob.data, C.int(blob.size))
-			C.duckdb_free(blob.data)
-		case C.DUCKDB_TYPE_TIMESTAMP:
-			val := (*[1 << 31]C.duckdb_timestamp)(unsafe.Pointer(colData))[r.cursor]
-			dst[i] = time.UnixMicro(int64(val.micros))
+	colCount := len(r.columns)
+
+	for colIdx := C.idx_t(0); colIdx < C.idx_t(colCount); colIdx++ {
+		vector := C.duckdb_data_chunk_get_vector(r.chunk, colIdx)
+		value, err := scanValue(vector, r.chunkRowIdx)
+		if err != nil {
+			return err
 		}
+		dst[colIdx] = value
 	}
 
-	r.cursor++
+	r.chunkRowIdx++
 
 	return nil
 }
 
+func get[T any](vector C.duckdb_vector, rowIdx C.idx_t) T {
+	ptr := C.duckdb_vector_get_data(vector)
+	xs := (*[1 << 31]T)(ptr)
+	return xs[rowIdx]
+}
+
+func decode(data any, out any) error {
+	return mapstructure.Decode(data, out)
+}
+
+func scanValue(vector C.duckdb_vector, rowIdx C.idx_t) (any, error) {
+	v, err := scan(vector, rowIdx)
+	if err != nil {
+		return nil, err
+	}
+
+	switch value := v.(type) {
+	case map[string]any, []any, Map, driver.Value:
+		return value, nil
+	case nil:
+		return nil, nil
+	default:
+		panic(fmt.Sprintf("BUG: found unexpected type when scanning: %T", value))
+	}
+}
+
+func scan(vector C.duckdb_vector, rowIdx C.idx_t) (any, error) {
+	validity := C.duckdb_vector_get_validity(vector)
+	if !C.duckdb_validity_row_is_valid(validity, rowIdx) {
+		return nil, nil
+	}
+
+	ty := C.duckdb_vector_get_column_type(vector)
+	typeId := C.duckdb_get_type_id(ty)
+	switch typeId {
+	case C.DUCKDB_TYPE_INVALID:
+		return nil, errInvalidType
+	case C.DUCKDB_TYPE_BOOLEAN:
+		return get[bool](vector, rowIdx), nil
+	case C.DUCKDB_TYPE_TINYINT:
+		return get[int8](vector, rowIdx), nil
+	case C.DUCKDB_TYPE_SMALLINT:
+		return get[int16](vector, rowIdx), nil
+	case C.DUCKDB_TYPE_INTEGER:
+		return int64(get[int32](vector, rowIdx)), nil
```

Comment:
This is purely because of the docs on `driver.Value`. Clearly this isn't enforced at all, but I'm unsure if `int32` would be handled properly.

// Value is a value that drivers must be able to handle.
// It is either nil, a type handled by a database driver's NamedValueChecker
// interface, or an instance of one of these types:
//
//   int64
//   float64
//   bool
//   []byte
//   string
//   time.Time
//

Reply:
Thanks, makes sense.